packages/react-native-lockdown/src/index.js

```diff
@@ -0,0 +1,127 @@
+const { warn } = console
+const { assign } = Object
+
+const path = require('node:path')
+
+const defaultGetRunModuleStatement = (moduleId) => `__r(${moduleId})`
+
+const ENTRY_FILE_MODULE_ID = 0
+
+const warnAboutAnomalies = () => {
+  const deferredWarnings = {
+    calledOnceOnly:
+      'LavaMoat: getRunModuleStatement was only called once instead of at least twice in quick succession, with react-native InitializeCore first and bundle entry second. This is suspicious. Lockdown might not be in effect.',
+    neverCalledWithEntryModule:
+      'LavaMoat: getRunModuleStatement was not called with moduleId 0. Lockdown was not applied.',
+  }
+
+  // Emit warnings only after Metro finishes bundling
+  process.on('exit', () => {
+    Object.values(deferredWarnings).forEach((message) => {
+      if (message) warn(message)
+    })
+  })
+
+  return function inspectCallLog(callLog) {
+    if (callLog.length > 1) {
+      deferredWarnings['calledOnceOnly'] = false
+    }
+
+    if (callLog.includes(ENTRY_FILE_MODULE_ID)) {
+      deferredWarnings['neverCalledWithEntryModule'] = false
+    }
+
+    if (callLog[0] === ENTRY_FILE_MODULE_ID) {
+      warn(
+        'LavaMoat: getRunModuleStatement was called with moduleId 0 first. That should represent the bundle entrypoint but react-native InitializeCore should run first. This is suspicious. Lockdown might not be in effect or might be breaking react-native core polyfills.'
+      )
+    }
+
+    if (new Set(callLog).size !== callLog.length) {
+      warn(
+        `LavaMoat: getRunModuleStatement was called with the same moduleId multiple times. It might be because some build is running twice with the same config instance. [${callLog.join(', ')}]`
+      )
+    }
+  }
+}
+
+module.exports = {
+  // lockdownSerializer
+  lockdownSerializer: ({ hermesRuntime = true } = {}, userConfig) => {
+    if (userConfig.getRunModuleStatement) {
+      warn(
+        'LavaMoat: You are using getRunModuleStatement in serializer config. Lavamoat will attempt to wrap it without breaking it, but if you are doing something unusual, we might affect how it works.'
+      )
+    }
+
+    const config = assign({}, userConfig)
+
+    const callLog = []
+
+    const previousGetRunModuleStatement =
+      userConfig.getRunModuleStatement ?? defaultGetRunModuleStatement
+
+    const inspectCallLog = warnAboutAnomalies()
+
+    // getRunModuleStatement
+    config.getRunModuleStatement = (moduleId) => {
+      callLog.push(moduleId)
+      inspectCallLog(callLog)
+      const runModuleStatement = previousGetRunModuleStatement(moduleId)
+
+      if (moduleId === ENTRY_FILE_MODULE_ID) {
+        return `hardenIntrinsics();${runModuleStatement}`
+      }
+
+      return runModuleStatement
+    }
+
+    if (!config.getPolyfills) {
+      config.getPolyfills = require('@react-native/js-polyfills')
+    } else if (typeof config.getPolyfills !== 'function') {
+      throw new Error('serializer.getPolyfills must be a function')
```

Comment:
```suggestion
      throw TypeError('Invalid options: getPolyfills must be a function')
```

- `new` is not necessary when used with intrinsic `Error`s
- `serializer` doesn't seem to reference anything here

Reply:
Also: if the user can provide their own function, is `@react-native/js-polyfills` an _optional_ peer dependency?

Reply:
updated ^

Reply:
nope not an optional peer dependency (which would be [peerDependenciesMeta](https://docs.npmjs.com/cli/v7/configuring-npm/package-json#peerdependenciesmeta) i think)

Reply:
but listed in our package.json as under `peerDependencies` and mentioned in ReadMe